tool/gravity/cli/ops.go

```diff
@@ -81,33 +82,87 @@ func uploadUpdate(ctx context.Context, tarballEnv *localenv.TarballEnvironment,
 		return trace.Wrap(err)
 	}
 
-	appPackage, err := install.GetAppPackage(tarballEnv.Apps)
+	application, err := install.GetApp(tarballEnv.Apps)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	env.PrintStep("Importing application %v v%v", appPackage.Name, appPackage.Version)
-	_, err = appservice.PullApp(appservice.AppPullRequest{
-		SrcPack: tarballEnv.Packages,
-		SrcApp:  tarballEnv.Apps,
-		DstPack: clusterPackages,
-		DstApp:  clusterApps,
-		Package: *appPackage,
+	var registries []string
+	err = utils.Retry(defaults.RetryInterval, defaults.RetryLessAttempts, func() error {
+		registries, err = getRegistries(ctx, env, cluster.ClusterState.Servers)
+		return trace.Wrap(err)
 	})
 	if err != nil {
-		if !trace.IsAlreadyExists(err) {
+		return trace.Wrap(err)
+	}
+	if len(registries) == 0 {
+		return trace.NotFound("could not determine cluster registry addresses")
+	}
+	images, err := docker.NewClusterImageService(registries[0])
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	sourcePackages := tarballEnv.Packages
+	sourceApps := tarballEnv.Apps
+
+	// Before importing check if the new version is an incremental upgrade
+	// and if so, check whether the currently installed version can be
+	// upgraded to it.
+	upgradeFrom, err := application.LabelAsLocator(pack.UpgradeFromLabel)
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+	if upgradeFrom != nil {
+		if !cluster.App.Package.IsEqualTo(*upgradeFrom) && !force {
+			return trace.BadParameter(`This cluster image was built as an incremental upgrade from %v.
```

Comment:
This does not state what the problem is - I'd be puzzled at what is expected.
`This cluster image was built as an incremental upgrade from %v which differs from the currently installed cluster image %v. Upgrading over a different image might break the cluster ...`